voting-dapp-template/move/sources/voting.move

```diff
@@ -0,0 +1,232 @@
+        );
+        let user_obj_signer = object::generate_signer(user_obj_constructor_ref);
+        move_to(&user_obj_signer, Vote {
+            voter: signer::address_of(sender),
+            vote,
+            amount,
+        });
+    }
+
+    // ======================== Read Functions ========================
+    public fun current_proposal(): (
+        Option<String>,
+        address,
+        u64,
+        u64,
+        u64,
+        u64
+    ) acquires ProposalRegistry{
+        let proposal = get_proposal_from_registry();
+        (
+            proposal.name,
+            proposal.creator,
+            proposal.start_time,
+            proposal.end_time,
+            proposal.yes_votes,
+            proposal.no_votes
+        )
+    }
+
+    public fun has_proposal_ended(): bool acquires ProposalRegistry {
+        let proposal = get_proposal_from_registry();
+        let curr = timestamp::now_seconds();
+        curr > proposal.end_time
+    }
+
+    public fun current_proposal_result(): (bool, u64, u64) acquires ProposalRegistry {
+        let proposal = get_proposal_from_registry();
+        (proposal.yes_votes > proposal.no_votes, proposal.yes_votes, proposal.no_votes)
+    }
+
+    public fun get_vote_obj(sender: address, proposal_registry_length: u64): Object<Vote> {
+        let seed = construct_object_seed(sender, proposal_registry_length);
+        object::address_to_object(object::create_object_address(&sender, seed))
+    }
+
+    public fun get_vote_obj_addr(sender: address, proposal_registry_length: u64): address {
+        let seed = construct_object_seed(sender, proposal_registry_length);
+        object::create_object_address(&sender, seed)
+    }
+
+    public fun get_vote(vote_obj: Object<Vote>): (address, bool, u64) acquires Vote {
+        let vote = borrow_global<Vote>(object::object_address(&vote_obj));
+        (vote.voter, vote.vote, vote.amount)
+    }
+
+    // ======================== Helper Functions ========================
+    // reduce overhead with a function call, performance optimization, gas etc.
+    inline fun get_proposal_from_registry(): Proposal acquires ProposalRegistry {
+        let proposal_registry = borrow_global<ProposalRegistry>(@voting_addr);
+        assert!(vector::length(&proposal_registry.proposals) > 0, E_PROPOSAL_DOES_NOT_EXIST);
+        *vector::borrow(&proposal_registry.proposals, vector::length(&proposal_registry.proposals) - 1)
+    }
+
+    fun construct_object_seed(sender: address, proposal_registry_length: u64 ): vector<u8> {
+        bcs::to_bytes(&string_utils::format3(&b"{}_user_{}_proposal_{}",
+            @voting_addr,
+            sender,
+            proposal_registry_length
+            )
+        )
+    }
+
+    // ======================== Unit Tests ========================
+    #[test_only]
+    use std::string;
+
+    #[test(aptos_framework = @std, owner = @voting_addr, alice = @0x1234)]
```

Comment:
usually I prefer to have test in its own module, you can refer to the [test in friend tech](https://github.com/aptos-labs/create-aptos-dapp/blob/main/examples/aptos-friend/move/tests/test_end_to_end.move)